docs/en/sql-reference/data-types/variant.md

```diff
@@ -272,3 +272,103 @@ $$)
 │ [1,2,3]             │ ᴺᵁᴸᴸ          │ ᴺᵁᴸᴸ │  ᴺᵁᴸᴸ │                ᴺᵁᴸᴸ │ [1,2,3] │
 └─────────────────────┴───────────────┴──────┴───────┴─────────────────────┴─────────┘
 ```
+
+
+## Comparing values of Variant type
+
+Values of a `Variant` type can be compared only with values with the same `Variant` type.
+
+The result of operator `<` for values `v1` with underlying type `T1` and `v2` with underlying type `T2`  of a type `Variant(..., T1, ... T2, ...)` is defined as follows:
+- If `T1 = T2 = T`, the result will be `v1.T < v2.T` (underlying values will be compared).
+- If `T1 != T2`, the result will be `T1 < T2` (type names will be compared).
```

Comment:
It can be so confusing that `(100 :: Int32) < (1 :: UInt32)`:

```sql
:) create table t(x Variant(UInt32, Int64)) ENGINE = Memory;

Ok.

0 rows in set. Elapsed: 0.015 sec. 

:) insert into t VALUES (1),(101-1),(100);

Ok.

3 rows in set. Elapsed: 0.003 sec. 

:) select * from t ORDER BY x;

┌─x───┐
│ 100 │
│ 1   │
│ 100 │
└─────┘

3 rows in set. Elapsed: 0.003 sec. 
```

I see that we just want to have a total ordering on `Variant`, but maybe we can keep cast to Variant monotonic:

```SQL

SELECT
    CAST('100', 'Int32') AS a,
    CAST('1', 'UInt32') AS b,
    a < b,
    CAST(a, 'Variant(UInt32, Int32)') < CAST(b, 'Variant(UInt32, Int32)')

Row 1:
──────
a:                                                                                                               100
b:                                                                                                               1
less(CAST('100', 'Int32'), CAST('1', 'UInt32')):                                                                 0
less(CAST(CAST('100', 'Int32'), 'Variant(UInt32, Int32)'), CAST(CAST('1', 'UInt32'), 'Variant(UInt32, Int32)')): 1
```

For example, we can Imagine situation when user has `Variant(T1, T2)` (e.g. `Variant(UInt32, Int32)`) but then wants to convert it to `leastSupertype(T1, T2)` (e.g. Int64) (or vise versa) and result ordering will be changed.

Reply:
> I see that we just want to have a total ordering on Variant, but maybe we can keep cast to Variant monotonic:

Yes, having total  ordering is the only reason. In general case Variant can have any data types inside that can be not comparable between each other, and I am not sure it is good idea to add extra logic for comparing between some subset of types. And actually, it is not trivial to implement as we can have a lot of different combinations of types that can be compared, like all integers, floats and decimals, dates and datetimes, arrays of integers, etc. And all this logic should be implemented in `ColumnVariant::compareAt()` where we don't have the data types of variants but only columns (so we even cannot use implementation of function `less`).

So, I decided to use the simpliest comparing rule that will allow to have total ordering

Reply:
And the user always can use specific variants in order by to make desired order, or cast the whole Variant to some common type:
```sql
avogar-dev :) select * from t ORDER BY if(variantType(x) == 'UInt32', x.UInt32, x.Int64);

SELECT *
FROM t
ORDER BY if(variantType(x) = 'UInt32', x.UInt32, x.Int64) ASC

Query id: 45153e4b-75e1-4363-b611-3dbf4a145e8d

┌─x───┐
│ 1   │
│ 100 │
│ 100 │
└─────┘

3 rows in set. Elapsed: 0.027 sec.

avogar-dev :) select * from t ORDER BY x::Int64;

SELECT *
FROM t
ORDER BY CAST(x, 'Int64') ASC

Query id: 6d987031-8aec-4bed-b54c-bb3a0d3dca57

┌─x───┐
│ 1   │
│ 100 │
│ 100 │
└─────┘

```

Reply:
I would keep current simple comparison as I don't see how we can properly implement something more advanced. Maybe we can improve it later as this type is still experimental and we can change its behaviour

Reply:
Ok, let's just add a warning to doc then? Also we may just forbid such variants that contains several different numeric types (e.g. both UInt32 and Int64) and allow it under setting similarly to `allow_suspicious_lowcardinality_types`.

Reply:
> Ok, let's just add a warning to doc then?

Sure, I will add it.

> Also we may just forbid such variants that contains several different numeric types (e.g. both UInt32 and Int64) and allow it under setting similarly to allow_suspicious_lowcardinality_types.

Sounds like a good idea, even maybe not only for different numeric types , but more general - types that have common supertype.
I see only 1 problem, we will have the same problem with Dynamic type that I am working on right now, and there we won't be able to check for similar types, because initially we will have 0 variants and will extend them dynamically, so it may happen that 1 part will have Dynamic with inner type UInt32 and 2 part with Int64, and when we will read from such table with ORDER BY we will have the same problem.
But Dynamic is abother story, I think we can add such setting for Variant, I will add it in this PR